### ISSUE TITLE
Misleading error message when the bucket name is not valid

### ISSUE THREAD

**Issue by pmontrasio** [jorgebastida/gordon#24]
I typed an invalid bucket name (with an upper case letter) and this is the output of gordon:

```
 $ gordon apply
Applying project...
  0001_p.json (cloudformation)
    UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS waiting... |                       
  0002_pr_r.json (custom)
Traceback (most recent call last):
  File "/home/me/.local/bin/gordon", line 11, in <module>
    sys.exit(main())
  File "/home/me/.local/lib/python2.7/site-packages/gordon/bin.py", line 97, in main
    getattr(obj, options.func)()
  File "/home/me/.local/lib/python2.7/site-packages/gordon/core.py", line 432, in apply
    getattr(self, 'apply_{}_template'.format(template_type))(name, filename, context)
  File "/home/me/.local/lib/python2.7/site-packages/gordon/core.py", line 476, in apply_custom_template
    outputs = template.apply(context, self)
  File "/home/me/.local/lib/python2.7/site-packages/gordon/actions.py", line 131, in apply
    action_outputs[action.name] = action.apply(context, project)
  File "/home/me/.local/lib/python2.7/site-packages/gordon/actions.py", line 217, in apply
    obj.upload_file(self.filename, ExtraArgs={'Metadata': {'sha1': file_hash}})
  File "/home/me/.local/lib/python2.7/site-packages/boto3/s3/inject.py", line 161, in object_upload_file
    ExtraArgs=ExtraArgs, Callback=Callback, Config=Config)
  File "/home/me/.local/lib/python2.7/site-packages/boto3/s3/inject.py", line 84, in upload_file
    extra_args=ExtraArgs, callback=Callback)
  File "/home/me/.local/lib/python2.7/site-packages/boto3/s3/transfer.py", line 641, in upload_file
    self._put_object(filename, bucket, key, callback, extra_args)
  File "/home/me/.local/lib/python2.7/site-packages/boto3/s3/transfer.py", line 651, in _put_object
    **extra_args)
  File "/home/me/.local/lib/python2.7/site-packages/botocore/client.py", line 258, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/home/me/.local/lib/python2.7/site-packages/botocore/client.py", line 548, in _make_api_call
    raise ClientError(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (NoSuchBucket) when calling the PutObject operation: The specified bucket does not exist
```

I expected an error on the creation of the bucket. I realized what happened when I tried to manually create the bucket: the AWS console told me the name was invalid. Apparently gordon silently ignored the error on creation and failed only on the first access to the bucket.

**Comment by pmontrasio** [comment]
Furthermore I suggest to leave the bucket name out of the configuration. It's not important and it could be randomly generated. If given a prefix such as ```gordon-<stage>-``` then it would be easy to spot it in the S3 console. It could be stored into a dot file in the project folder for easier retrieval, maybe ```private.conf```

**Comment by jorgebastida** [comment]
I think there are two things we can do to make the experience better / not broken.

The most obvious one is to make gordon not proceed applying templates if the previous didn't finish successfully.

The second thing which would make gordon better would be to add a validation process before applying templates. In this case validate that the name of the bucket is correct... but there are many more situations where gordon could stop people from making configuration mistakes.

About adding a "state" configuration file... that is something which crossed my mind in the past, but the problem is that you would need to have one per stage and region... which might end looking like a mess. Will give it a second think.

**Comment by pmontrasio** [comment]
First of all, maybe we should split this issue in two, one about the error message and one about the configuration. Sorry for not doing it myself right away.

About the former: if AWS returns an error (does it?) gordon should just print it and quit.

About the latter: I'm not particularly familiar with AWS and I believe that if gordon and the other similar tools will get popular it will be because many other AWS newbies are using them. So, we don't know much about the reasons why some AWS services are tied together and we don't care much. I'm perfectly OK with gordon doing it's magic and hiding everything to me (but be careful about decisions that could cost money to people). Creating the bucket and immediately destroying it would be fine, or uploading the code as a ZIP file. Having to handle too many details would slow me down and create bugs. An example with gordon: I still don't know why I couldn't get the API gateway to pass my parameters to my ```event``` object but I'm probably misconfiguring something.  

An example from the past: many people started using Rails 10 years ago without knowing anything about Ruby. Ruby was just the scripting language of Rails, used to fill in the blank of the generated scaffolds. It got pretty successful anyway and most of the Ruby developers nowadays develop with Rails. Replace Ruby with AWS and Rails with gordon and you see what I mean. Then you need only to fill in some JS code in the template generated for the code.js file (or Python).

**Comment by jorgebastida** [comment]
Just fixed the first bit of this issue. It is ready on master, and will be released as part of ``0.2.1`` https://github.com/jorgebastida/gordon/commit/a8e9f2723e69208c49236ccf35faaba2616c26af

The tl;dr version is that if applying a template fails, gordon will tell you so and exit.

About the second issue, I've added validation for the stage name and the ``code-bucket`` setting but... yes, next versions should aim to remove complexity (like the parameters comment you mentioned in the other PR) as well as trying to warn users when some configurations are obviously wrong.

These changes are on master and will get released as part of ``0.2.1``. Thanks!
